test/xds/xds_server_certificate_providers_test.go

```diff
@@ -0,0 +1,477 @@
+	})
+	if err != nil {
+		t.Fatalf("Failed to create bootstrap configuration: %v", err)
+	}
+
+	// Configure xDS credentials to be used on the server-side.
+	creds, err := xdscreds.NewServerCredentials(xdscreds.ServerOptions{FallbackCreds: insecure.NewCredentials()})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Initialize an xDS-enabled gRPC server and register the stubServer on it.
+	// Pass it a mode change server option that pushes on a channel the mode
+	// changes to "not serving".
+	modeCh := make(chan struct{})
+	modeChangeOpt := xds.ServingModeCallback(func(addr net.Addr, args xds.ServingModeChangeArgs) {
+		if args.Mode == connectivity.ServingModeServing {
+			close(modeCh)
+		}
+	})
+	server, err := xds.NewGRPCServer(grpc.Creds(creds), modeChangeOpt, xds.BootstrapContentsForTesting(bs))
+	if err != nil {
+		t.Fatalf("Failed to create an xDS enabled gRPC server: %v", err)
+	}
+	testgrpc.RegisterTestServiceServer(server, &testService{})
+	defer server.Stop()
+
+	// Create a local listener and pass it to Serve().
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("testutils.LocalTCPListener() failed: %v", err)
+	}
+
+	go func() {
+		if err := server.Serve(lis); err != nil {
+			t.Errorf("Serve() failed: %v", err)
+		}
+	}()
+
+	// Create an inbound xDS listener resource for the server side that contains
+	// mTLS security configuration. Since the received certificate provider
+	// instance name would be missing in the bootstrap configuration, this
+	// resource is expected to NACKed by the xDS client.
+	host, port, err := hostPortFromListener(lis)
+	if err != nil {
+		t.Fatalf("Failed to retrieve host and port of server: %v", err)
+	}
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultServerListener(host, port, e2e.SecurityLevelMTLS, "routeName")},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the NACK from the xDS client.
+	select {
+	case <-nackCh:
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for an NACK from the xDS client for the LDS response")
+	}
+
+	// Wait a short duration and ensure that the server does not enter ""
+	select {
+	case <-time.After(2 * defaultTestShortTimeout):
+	case <-modeCh:
+		t.Fatal("Server changed to serving mode when not expected to")
+	}
+
+	// Create a client that uses insecure creds and verifiy that RPCs don't
+	// succeed.
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	client := testgrpc.NewTestServiceClient(cc)
+	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+	if _, err := client.EmptyCall(sCtx, &testpb.Empty{}); status.Code(err) != codes.DeadlineExceeded {
+		t.Fatalf("EmptyCall() failed: %v", err)
+	}
```

Comment:
I should use this check for not serving (i.e. Accept() and Close() server side). Seems to differ from Doug's description: "If the connection is gracefully closed (and no new connection can be made) you'll just get UNAVAILABLE + connection refused or no addresses or failed to handshake or something" for graceful close than RPC.

Reply:
Actually, thinking about this more, is there a way to poll until you get a more specific certain error type (and not just set a small deadline and expect it to fail)? I don't know if there is but just wondering

Reply:
*If* the connection was established (i.e, the client channel went to `READY`), and then breaks, it will lead to the client channel transitioning through `IDLE` to `TRANSIENT_FAILURE` (when you can no longer establish a connection to the server). And in that case, yes, the RPC will fail with `UNAVAILABLE`.

But in this case, the connection never fails because there is no `Accept` followed by `Close`. The connection attempt is simply blocked. This means that the client channel will be in `CONNECTING`. And therefore the picker will return `ErrNoSubConnsAvailable` when an RPC is attempted, and the only way this RPC can fail is when the deadline expires (since we don't expect a picker update on the client).

When you make your change to do `Accept` followed by `Close`, you should be able to use the top-level test context instead of a context with a short deadline while making the RPC and check for `UNAVAILABLE` and probably even an error string. But unfortunately, that is not possible today.

Reply:
Ohhhh right, I was thinking as if my change was already in master with respect to Accept() + Close(), and not blocking as it is now. I'll probably switch this over (and I have tests for this Accept() + Close() - I thought hard about the invariant of a signal that a connection was Accepted + Closed).